### PR TITLE
[UNR-3159] Make singletons work with zoning

### PR DESCRIPTION
**Pull request by alastairdglennie**
[4.23 PR] https://github.com/improbableio/UnrealEngine/pull/431
[4.24 PR] https://github.com/improbableio/UnrealEngine/pull/430

Ticket prompting work: [Singletons don't work correctly with Zoning](https://improbableio.atlassian.net/browse/UNR-3159)
Related ticket: [[Spike] Reconsider singletons in the Unreal GDK](https://improbableio.atlassian.net/browse/UNR-2333)

#### Description
GSM doesn't handle singletons anymore and SingletonManager component is deleted.
Singletons still exist but much of their lifecycle, i.e. linking on non-auith servers, is handled in the same way as stably-named Actors.

The class specifier still exists so users can declare which Actors should exist once in the world (and should have their stable name set).

The SpatialOS component still exists so singletons can be added to the AlwaysRelefvant component set.

A consequence of this work is fixing the linked Jira ticket relating to zoning where singleton entities could previously never spawn if the lb strat strategies that the singleton should be on a different worker to the GSM.

#### Tests
Tested in multiple test gyms
Currently building in the example project, however, there are multiple other bugs related to run this with zoning.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -611,7 +611,7 @@ int64 USpatialActorChannel::ReplicateActor()
 
 			// If we're not offloading AND either load balancing isn't enabled or it is and we're spawning an Actor that we know
 			// will be load-balanced to another worker then preemptively set the role to SimulatedProxy.
-			if (!USpatialStatics::IsSpatialOffloadingEnabled() && (!SpatialGDKSettings->bEnableUnrealLoadBalancer || !NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor)))
+			if (!USpatialStatics::IsSpatialOffloadingEnabled() && !(SpatialGDKSettings->bEnableUnrealLoadBalancer && NetDriver->LoadBalanceStrategy->ShouldHaveAuthority(*Actor)))
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:614]: clearer imo

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -1105,12 +1105,6 @@ bool USpatialActorChannel::TryResolveActor()
 		return false;
 	}
 
-	// If a Singleton was created, update the GSM with the proper Id.
-	if (Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
-	{
-		NetDriver->GlobalStateManager->UpdateSingletonEntityId(Actor->GetClass()->GetPathName(), EntityId);
-	}
-
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp:1113]: gsm isn't used to track singletons anymore, we can delete this call

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp [hunk 3]
```diff
@@ -40,7 +40,6 @@ void FSpatialNetBitReader::DeserializeObjectRef(FUnrealObjectRef& ObjectRef)
 	}
 
 	SerializeBits(&ObjectRef.bNoLoadOnClient, 1);
-	SerializeBits(&ObjectRef.bUseSingletonClassPath, 1);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp:43]: deleted this component property

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitWriter.cpp [hunk 4]
```diff
@@ -37,7 +37,6 @@ void FSpatialNetBitWriter::SerializeObjectRef(FUnrealObjectRef& ObjectRef)
 	}
 
 	SerializeBits(&ObjectRef.bNoLoadOnClient, 1);
-	SerializeBits(&ObjectRef.bUseSingletonClassPath, 1);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitWriter.cpp:40]: deleted this component property

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 5]
```diff
@@ -365,7 +365,7 @@ void USpatialNetDriver::CreateAndInitializeCoreClasses()
 
 	const USpatialGDKSettings* SpatialSettings = GetDefault<USpatialGDKSettings>();
 #if !UE_BUILD_SHIPPING
-	// If metrics display is enabled, spawn a singleton actor to replicate the information to each client
+	// If metrics display is enabled, spawn an Actor to replicate the information to each client.
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:368]:
```suggestion
	// If metrics display is enabled, spawn a singleton Actor to replicate the information to each client.
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 6]
```diff
@@ -709,11 +708,11 @@ void USpatialNetDriver::OnLevelAddedToWorld(ULevel* LoadedLevel, UWorld* OwningW
...
 		// otherwise, load balancing is enabled, so check the lb strategy.
-		if (Actor->GetIsReplicated() &&
-			(!bLoadBalancingEnabled || LoadBalanceStrategy->ShouldHaveAuthority(*Actor)))
+		if (Actor->GetIsReplicated())
 		{
-			Actor->Role = ROLE_Authority;
-			Actor->RemoteRole = ROLE_SimulatedProxy;
+			const bool bRoleAuthoritative = !bLoadBalancingEnabled || LoadBalanceStrategy->ShouldHaveAuthority(*Actor);
+			Actor->Role = bRoleAuthoritative ? ROLE_Authority : ROLE_SimulatedProxy;
+			Actor->RemoteRole = bRoleAuthoritative ? ROLE_SimulatedProxy : ROLE_Authority;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:715]: Previously this only functioned correctly if elsewhere in the codebase we set roles to simulated proxy first, we can remove that dependency by setting to simulated proxy here if not set authority

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 7]
```diff
@@ -871,14 +870,6 @@ void USpatialNetDriver::NotifyActorDestroyed(AActor* ThisActor, bool IsSeamlessT
...
 
 	const bool bIsServer = ServerConnection == nullptr;
-
-	// Remove the record of destroyed singletons.
-	if (ThisActor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
-	{
-		// We check for this not being a server below to make sure we don't call this incorrectly in single process PIE sessions.
-		GlobalStateManager->RemoveSingletonInstance(ThisActor);
-	}
-
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:881]: GSM doesn't care about singletons anymore

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 8]
```diff
@@ -1346,12 +1337,7 @@ void USpatialNetDriver::ServerReplicateActors_ProcessPrioritizedActors(UNetConne
 						continue;
 					}
 
-					if (!Actor->HasAuthority())
-					{
-						// Trying to replicate Actor which we don't have authority over.
-						// Remove after UNR-961
-						continue;
-					}
+					check(Actor->HasAuthority());
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:1340]: driveby, jira ticket was done last year lol

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 9]
```diff
@@ -2307,25 +2293,24 @@ USpatialActorChannel* USpatialNetDriver::CreateSpatialActorChannel(AActor* Actor
 	check(NetConnection != nullptr);
 
 	USpatialActorChannel* Channel = nullptr;
-	if (Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
+
+	if (Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton) && !LoadBalanceStrategy->ShouldHaveAuthority(*Actor))
 	{
-		Channel = GlobalStateManager->AddSingleton(Actor);
+		return Channel;
+	}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2300]: Currently, for complicated reasons, singletons initially have authority on every server initially. We need to explicitly prevent each server trying to creating an entity for them. Previously this was handled within GSM::AddSingleton (and based on GSM authority), now there's a static method in the Singleton component file (seemed best place vs add MORE to the netdriver)

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 10]
```diff
@@ -2307,25 +2293,24 @@ USpatialActorChannel* USpatialNetDriver::CreateSpatialActorChannel(AActor* Actor
...
-		{
 #if ENGINE_MINOR_VERSION <= 22
-			Channel->SetChannelActor(Actor);
+		Channel->SetChannelActor(Actor);
 #else
-			Channel->SetChannelActor(Actor, ESetChannelActorFlags::None);
+		Channel->SetChannelActor(Actor, ESetChannelActorFlags::None);
 #endif
-		}
 	}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2314]: this looks messy, but it's just a deindentation from being coupled with the if statement that my comment above refers to

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 11]
```diff
@@ -1110,14 +1100,6 @@ AActor* USpatialReceiver::CreateActor(UnrealMetadata* UnrealMetadataComp, SpawnD
...
 	}
 
-	const bool bIsServer = NetDriver->IsServer();
-
-	// Initial Singleton Actor replication is handled with GlobalStateManager::LinkExistingSingletonActors
-	if (bIsServer && ActorClass->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
-	{
-		return FindSingletonActor(ActorClass);
-	}
-
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:1120]: Instead, singletons now follow the StablyNamedActor logic from the `TryGetOrCreateActor` method above

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 12]
```diff
@@ -2174,16 +2138,7 @@ void USpatialReceiver::ResolvePendingOperations(UObject* Object, const FUnrealOb
...
-	if (Object->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton) && !Object->IsFullNameStableForNetworking())
-	{
-		// When resolving a singleton, also resolve using class path (in case any properties
-		// were set from a server that hasn't resolved the singleton yet)
-		FUnrealObjectRef ClassObjectRef = FUnrealObjectRef::GetSingletonClassRef(Object, PackageMap);
-		if (ClassObjectRef.IsValid())
-		{
-			ResolveIncomingOperations(Object, ClassObjectRef);
-		}
-	}
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2186]: this feels redundant now, need someone to reason with me whether that's true or not

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSnapshotManager.cpp [hunk 13]
```diff
@@ -183,7 +183,7 @@ void SpatialSnapshotManager::LoadSnapshot(const FString& SnapshotName)
 			// Check if this is the GSM
 			for (auto& ComponentData : EntityToSpawn)
 			{
-				if (ComponentData.component_id == SpatialConstants::SINGLETON_MANAGER_COMPONENT_ID)
+				if (ComponentData.component_id == SpatialConstants::STARTUP_ACTOR_MANAGER_COMPONENT_ID)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSnapshotManager.cpp:186]: SIngletonManager component deleted so use StartupActorManager to identify gsm

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialStaticComponentView.cpp [hunk 14]
```diff
@@ -70,9 +69,6 @@ void USpatialStaticComponentView::OnAddComponent(const Worker_AddComponentOp& Op
 	case SpatialConstants::SPAWN_DATA_COMPONENT_ID:
 		Data = MakeUnique<SpatialGDK::SpawnData>(Op.data);
 		break;
-	case SpatialConstants::SINGLETON_COMPONENT_ID:
-		Data = MakeUnique<SpatialGDK::Singleton>(Op.data);
-		break;
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialStaticComponentView.cpp:75]: We have 0 use for this component within workers, it's just used to make singletons always relevant, saw no need to store this here

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp [hunk 15]
```diff
@@ -220,7 +220,7 @@ TArray<FWorkerComponentData> EntityFactory::CreateEntityComponents(USpatialActor
 	// Actors with bNetLoadOnClient=false also need a StablyNamedObjectRef for linking in the case of loading from a snapshot or the server crashes and restarts.
 	TSchemaOption<FUnrealObjectRef> StablyNamedObjectRef;
 	TSchemaOption<bool> bNetStartup;
-	if (Actor->HasAnyFlags(RF_WasLoaded) || Actor->bNetStartup)
+	if (Actor->HasAnyFlags(RF_WasLoaded) || Actor->bNetStartup || Class->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:223]: In order for singletons to be linked in the same way as startup Actors without all the GSM plumbing, we need to make sure we set them as stably named

File: SpatialGDK/Source/SpatialGDK/Public/Interop/GlobalStateManager.h [hunk 16]
```diff
@@ -101,7 +87,7 @@ class SPATIALGDK_API UGlobalStateManager : public UObject
 private:
 	void SetDeploymentMapURL(const FString& MapURL);
 	void SendSessionIdUpdate();
-	void LinkExistingSingletonActor(const UClass* SingletonClass);
+	//void LinkExistingSingletonActor(const UClass* SingletonClass);
```
Comment (ImprobableNic) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/GlobalStateManager.h:90]: Nit : Leftover comment

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp (hunk 15 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:223]: How are we sure that the singleton's outer name is also stable ?

**Comment by cmsmithio** [comment]
overall lgtm.  i'll pull this down and test against NWX to get you some more data

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp (hunk 15 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:223]: I'm not sure we are, though I think we should maybe just have that as a requirement of using singletons in the doc, i.e. you must set the outer as something which'll live as long as a singleton, so another singleton or leave it as NULL which'll be set to the persistent level? Not too familiar with how outer objects work so would need some input from others

**Comment by alastairdglennie** [comment]
Need to bump require setup version

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp (hunk 15 above)
Comment (ImprobableNic) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:223]: Was looking at how we spawned the spatial debugger, which was done through GetWorld()->SpawnActor<>(). That would create a dynamic actor, right ?

File: SpatialGDK/Source/SpatialGDK/Public/Interop/GlobalStateManager.h (hunk 16 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/GlobalStateManager.h:90]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 17]
```diff
@@ -2315,25 +2303,24 @@ USpatialActorChannel* USpatialNetDriver::CreateSpatialActorChannel(AActor* Actor
 	check(NetConnection != nullptr);
 
 	USpatialActorChannel* Channel = nullptr;
-	if (Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
+
+	if (Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton) && !SpatialGDK::Singleton::ShouldHaveLocalWorkerAuthority(Actor, LoadBalanceStrategy, StaticComponentView))
 	{
-		Channel = GlobalStateManager->AddSingleton(Actor);
+		return Channel;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2309]: Hmm, we still need a channel if you're non-auth. And doesn't CreateSpatialActorChannel get called in that instance?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 18]
```diff
@@ -2315,25 +2303,24 @@ USpatialActorChannel* USpatialNetDriver::CreateSpatialActorChannel(AActor* Actor
...
+	}
+
+	Channel = static_cast<USpatialActorChannel*>(NetConnection->CreateChannelByName(NAME_Actor, EChannelCreateFlags::OpenedLocally));
+	if (Channel == nullptr)
+	{
+		UE_LOG(LogSpatialOSNetDriver, Warning, TEXT("Failed to create a channel for actor %s."), *GetNameSafe(Actor));
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2315]: nit: could just return nullptr here, and simplify the rest of the func

File: SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp [hunk 19]
```diff
@@ -543,7 +319,7 @@ bool UGlobalStateManager::HandlesComponent(const Worker_ComponentId ComponentId)
 {
 	switch (ComponentId)
 	{
-		case SpatialConstants::SINGLETON_MANAGER_COMPONENT_ID:
+		//case SpatialConstants::SINGLETON_MANAGER_COMPONENT_ID:
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/GlobalStateManager.cpp:322]: 🔥

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 12 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2191]: This was because you could refer to a singleton using a stably named path (which was actually the class path), and the bool in the unreal obj ref. As you've removed that now, this isn't necessary.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp (hunk 15 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:223]: Yep - that'd be a dynamic actor @ImprobableNic 
Do we have a need for singletons anymore? I'm more of the opinion we should remove them entirely, and leave it up to the user.

File: SpatialGDK/Source/SpatialGDK/Public/Schema/Singleton.h [hunk 20]
```diff
@@ -2,32 +2,42 @@
...
 namespace SpatialGDK
 {
-
-struct Singleton : Component
-{
-	static const Worker_ComponentId ComponentId = SpatialConstants::SINGLETON_COMPONENT_ID;
-
-	Singleton() = default;
-	Singleton(const Worker_ComponentData& Data)
+	struct Singleton : Component
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Schema/Singleton.h:16]: nit: don't indent inside namespace

File: SpatialGDK/Extras/schema/core_types.schema [hunk 21]
```diff
@@ -12,8 +12,9 @@ type UnrealObjectRef {
     // a reference, e.g. anything inside streaming levels should not be loaded.
     option<bool> no_load_on_client = 4;
     option<UnrealObjectRef> outer = 5;
-    // Singleton objects can be referred to by their class path in the case of an
+    // Actors such as the game state, game mode and level script Actors (formerly
+    // known as Singletons) can be referred to by their class path in the case of an
     // authoritative server that hasn't checked the singleton entity yet. This bool
     // will differentiate that from their class pointer.
```
Comment (alastairdglennie) [inline SpatialGDK/Extras/schema/core_types.schema:18]:
```suggestion
    // authoritative server that hasn't checked the singleton entity yet. This bool
    // will indicate that non-auth servers should try to load such Actors from
    // their package map.
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 12 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2191]: Do we need this back now?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp [hunk 22]
```diff
@@ -256,29 +258,24 @@ TArray<FWorkerComponentData> EntityFactory::CreateEntityComponents(USpatialActor
 		ComponentDatas.Add(AuthorityIntent::CreateAuthorityIntentData(IntendedVirtualWorkerId));
 	}
 
-	if (NetDriver->SpatialDebugger != nullptr)
+#if !UE_BUILD_SHIPPING
+	if (SpatialSettings->SpatialDebugger != nullptr)
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:262]:
```suggestion
	if (NetDriver->SpatialDebugger != nullptr)
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp [hunk 23]
```diff
@@ -256,29 +258,24 @@ TArray<FWorkerComponentData> EntityFactory::CreateEntityComponents(USpatialActor
 		ComponentDatas.Add(AuthorityIntent::CreateAuthorityIntentData(IntendedVirtualWorkerId));
 	}
 
+#if !UE_BUILD_SHIPPING
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/EntityFactory.cpp:261]: We only spawn the SApatialDebugger in non-shipping, so add the SpatialDebugging mode when in shipping mode is redundant

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 24]
```diff
@@ -273,35 +273,30 @@ void ASpatialDebugger::OnEntityRemoved(const Worker_EntityId EntityId)
 
 void ASpatialDebugger::ActorAuthorityChanged(const Worker_AuthorityChangeOp& AuthOp) const
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:274]: just changed this to check for non-auth or non auth intent ops (receiver has a stricter conditional so this should happen), also reduces indentation in method

File: SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h [hunk 25]
```diff
@@ -85,7 +86,12 @@ struct SPATIALGDK_API FUnrealObjectRef
 	SpatialGDK::TSchemaOption<FString> Path;
 	SpatialGDK::TSchemaOption<FUnrealObjectRef> Outer;
 	bool bNoLoadOnClient = false;
-	bool bUseSingletonClassPath = false;
+	// If this field is set to true, we are saying that the Actor will exist at most once on the given worker.
+	// In addition, if we receive information for an Actor of the same class over the network, then this data
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h:90]:
```suggestion
	// In addition, if we receive information for an Actor of this class over the network, then this data
```

File: SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h [hunk 26]
```diff
@@ -85,7 +86,12 @@ struct SPATIALGDK_API FUnrealObjectRef
...
+	// should be applied to this Actor we've already spawned (where another worker created the entity). This
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Public/Schema/UnrealObjectRef.h:91]:
```suggestion
	// should be applied to the Actor we've already spawned (where another worker created the entity). This
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp (hunk 3 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp:43]: didn't this get added back in as `use_class_path_to_load_object`

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp:43]: ayay

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 12 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2191]: I believe so, because any object we're no longer referring too as singleton, could have two unrealobjrefs. One, for non-auth servers prior to knowing the entity id. And the other when we have the entity id.

File: SpatialGDK/Source/SpatialGDK/Private/Schema/UnrealObjectRef.cpp [hunk 27]
```diff
@@ -121,10 +128,11 @@ FUnrealObjectRef FUnrealObjectRef::FromObjectPtr(UObject* ObjectValue, USpatialP
 					}
 				}
 
-				// If this is a singleton that hasn't been resolved yet, send its class path instead.
-				if (ObjectValue->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
+				// If this is am Actor that should exist once per Worker (e.g. GameMode, GameState) and hasn't been resolved yet,
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Schema/UnrealObjectRef.cpp:131]:
```suggestion
				// If this is an Actor that should exist once per Worker (e.g. GameMode, GameState) and hasn't been resolved yet,
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 12 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2163]: Since deciding to get rid of them altogether, this should stay deleted

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp (hunk 17 above)
Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp:2294]: I think GitHub has misdirected this comment

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 28]
```diff
@@ -482,7 +477,9 @@ void USpatialReceiver::HandleActorAuthority(const Worker_AuthorityChangeOp& Op)
 		NetDriver->VirtualWorkerTranslationManager->AuthorityChanged(Op);
 	}
 
-	if (NetDriver->SpatialDebugger != nullptr)
+	if (NetDriver->SpatialDebugger != nullptr
+		&& Op.authority == WORKER_AUTHORITY_AUTHORITATIVE
+		&& Op.component_id == SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:482]: driveby?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 12 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2163]: But we're still using `use_class_path_to_load_object`?

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 24 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:274]: ah

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h [hunk 29]
```diff
@@ -129,10 +129,7 @@ inline void AddObjectRefToSchema(Schema_Object* Object, Schema_FieldId Id, const
 	{
 		AddObjectRefToSchema(ObjectRefObject, UNREAL_OBJECT_REF_OUTER_ID, *ObjectRef.Outer);
 	}
-	if (ObjectRef.bUseSingletonClassPath)
-	{
-		Schema_AddBool(ObjectRefObject, UNREAL_OBJECT_REF_USE_SINGLETON_CLASS_PATH_ID, ObjectRef.bUseSingletonClassPath);
-	}
+	Schema_AddBool(ObjectRefObject, UNREAL_OBJECT_REF_USE_CLASS_PATH_TO_LOAD_ID, ObjectRef.bUseClassPathToLoadObject);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h:132]: Why change this usage from an optional?

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h:132]: It didn't make sense being an optional to me

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h:132]: If it wasn't set it was considered false.
I see the schema itself is still an optional.
Done for bandwidth reasons I believe, do we know the tradeoff between an empty optional, and a bool? Same usage?

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h:132]: Oh I thought I changed the schema too. My opinion is that having a optional<bool> over just a bool for something infrequently changing just adds more ambiguity than potential bits saved per minute than it's worth? If option<bool> was 100% efficient than bool wouldn't we then always just use bool?

Comment (alastairdglennie) [reply SpatialGDK/Source/SpatialGDK/Public/Utils/SchemaUtils.h]: Reverted to being optional in schema for network benefit of sending nothing where the value is false. Kept as just a bool in the c++ class.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 30]
```diff
@@ -2177,16 +2135,18 @@ void USpatialReceiver::ResolvePendingOperations(UObject* Object, const FUnrealOb
 	UE_LOG(LogSpatialReceiver, Verbose, TEXT("Resolving pending object refs and RPCs which depend on object: %s %s."), *Object->GetName(), *ObjectRef.ToString());
 
 	ResolveIncomingOperations(Object, ObjectRef);
-	if (Object->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton) && !Object->IsFullNameStableForNetworking())
+
+	// When resolving a Actor that should uniquely exist in a deployment, e.g. GameMode, GameState, LevelScriptActors, we also
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp:2139]:
```suggestion
	// When resolving an Actor that should uniquely exist in a deployment, e.g. GameMode, GameState, LevelScriptActors, we also
```
